### PR TITLE
[prolly] filteredIter optimization for exact prefix ranges

### PR DESCRIPTION
**Pull request by max-hoffman**
Index range iteration uses a callback that is arbitrarily flexible but expensive. I changed index table access to only perform partial index scans for complete prefixes, and when the prefix fields equality conditions the generality of the index range callback is overkill. We just need to scan from the partial key `(field1, ..., fieldn, nil, ...)` to one higher than the partial key `(field1, fieldn+1, nil, ...)`.

This PR differentiates between `RangeField.StrictKey` and `.Equal` attributes to differentiate max-1-row and an equality restriction.

Still need to do follow-up tracing, but this is in response to the queries from TPC-C below. The string ones are much more common. Each of these use a set of equality filters than only partially completes a secondary index prefix. All of them spend ~5ms of CPU time executing `Range.Matches`, which is mostly eliminated with this PR.
```sql
SELECT o_entry_d FROM orders1  WHERE o_w_id = 1  AND o_d_id = 5  AND o_c_id = 1891  ORDER BY o_id DESC;
SELECT c_id  FROM customer1 WHERE c_w_id = 1 AND c_d_id= 6 AND c_last='ABLECALLYABLE' ORDER BY c_first;
SELECT o_id, o_carrier_id, o_entry_d FROM orders1 WHERE o_w_id = 1 AND o_d_id = 9 AND o_c_id = 1709 ORDER BY o_id DESC
```

### CODE REVIEW

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>c5ece8a</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>c5ece8a</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_p95 | to_latency_p95 | is_faster |
|---------------------|------------------|----------------|-----------|
| tpcc-scale-factor-1 | 81.48            | 81.48          | 0         |


|      test_name      | server_name |              server_version              |  tps  |      test_name      | server_name |              server_version              |  tps  | is_faster |
|---------------------|-------------|------------------------------------------|-------|---------------------|-------------|------------------------------------------|-------|-----------|
| tpcc-scale-factor-1 | dolt        | 850ec4bfb05923204e33bd45f9524d267f71dd05 | 26.13 | tpcc-scale-factor-1 | dolt        | c5ece8a7dce939b37a028bdf04f19a16ac8a0759 | 25.78 | 0         |

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>102c984</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>102c984</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 2.97                | 2.97              | 0         |
| groupby_scan         | 17.32               | 17.32             | 0         |
| index_join           | 5.37                | 5.47              | 0         |
| index_join_scan      | 2.22                | 2.22              | 0         |
| index_scan           | 54.83               | 53.85             | 0         |
| oltp_point_select    | 0.52                | 0.52              | 0         |
| oltp_read_only       | 8.28                | 8.43              | 0         |
| select_random_points | 0.83                | 0.84              | 0         |
| select_random_ranges | 0.97                | 0.97              | 0         |
| table_scan           | 55.82               | 55.82             | 0         |
| types_table_scan     | 139.85              | 139.85            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 6.21                | 6.21              | 0         |
| oltp_insert           | 3.02                | 3.07              | 0         |
| oltp_read_write       | 15.0                | 15.0              | 0         |
| oltp_update_index     | 3.19                | 3.19              | 0         |
| oltp_update_non_index | 3.07                | 3.07              | 0         |
| oltp_write_only       | 6.55                | 6.67              | 0         |
| types_delete_insert   | 6.79                | 6.79              | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 99.999360</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d791fc1</TD>
<TD>not ok</TD>
<TD>38</TD>
</TR>
<TR><TD>d791fc1</TD>
<TD>ok</TD>
<TD>5937419</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d791fc1</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.99936</TD>
</TR>
</table>

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 99.999360</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>9c2117e</TD>
<TD>not ok</TD>
<TD>38</TD>
</TR>
<TR><TD>9c2117e</TD>
<TD>ok</TD>
<TD>5937419</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>9c2117e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.99936</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_p95 | to_latency_p95 | is_faster |
|---------------------|------------------|----------------|-----------|
| tpcc-scale-factor-1 | 84.47            | 82.96          | 0         |


|      test_name      | server_name |              server_version              |  tps  |      test_name      | server_name |              server_version              | tps  | is_faster |
|---------------------|-------------|------------------------------------------|-------|---------------------|-------------|------------------------------------------|------|-----------|
| tpcc-scale-factor-1 | dolt        | 69d9d977ea11fbc865622cad83904aedab3a9c94 | 25.19 | tpcc-scale-factor-1 | dolt        | 9c2117e794bc1593081bb68e74c23bfcebd75513 | 27.9 | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 2.97                | 2.97              | 0         |
| groupby_scan         | 17.32               | 17.32             | 0         |
| index_join           | 5.37                | 5.37              | 0         |
| index_join_scan      | 2.22                | 2.22              | 0         |
| index_scan           | 52.89               | 54.83             | 0         |
| oltp_point_select    | 0.5                 | 0.52              | 0         |
| oltp_read_only       | 8.13                | 8.43              | 0         |
| select_random_points | 0.81                | 0.84              | 0         |
| select_random_ranges | 0.95                | 0.97              | 0         |
| table_scan           | 54.83               | 55.82             | 0         |
| types_table_scan     | 134.9               | 139.85            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 6.09                | 6.09              | 0         |
| oltp_insert           | 2.97                | 3.02              | 0         |
| oltp_read_write       | 14.73               | 15.0              | 0         |
| oltp_update_index     | 3.13                | 3.13              | 0         |
| oltp_update_non_index | 3.02                | 3.07              | 0         |
| oltp_write_only       | 6.55                | 6.67              | 0         |
| types_delete_insert   | 6.79                | 6.79              | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 3.02                | 2.97              | 0         |
| groupby_scan         | 17.32               | 17.32             | 0         |
| index_join           | 5.47                | 5.37              | 0         |
| index_join_scan      | 2.22                | 2.22              | 0         |
| index_scan           | 54.83               | 53.85             | 0         |
| oltp_point_select    | 0.52                | 0.52              | 0         |
| oltp_read_only       | 8.43                | 8.28              | 0         |
| select_random_points | 0.83                | 0.84              | 0         |
| select_random_ranges | 0.97                | 0.97              | 0         |
| table_scan           | 55.82               | 55.82             | 0         |
| types_table_scan     | 139.85              | 137.35            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 6.09                | 6.09              | 0         |
| oltp_insert           | 3.02                | 3.02              | 0         |
| oltp_read_write       | 15.0                | 15.0              | 0         |
| oltp_update_index     | 3.13                | 3.13              | 0         |
| oltp_update_non_index | 3.07                | 3.07              | 0         |
| oltp_write_only       | 6.55                | 6.67              | 0         |
| types_delete_insert   | 6.79                | 6.79              | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f3ab034</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f3ab034</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>a5917e0</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>a5917e0</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/index/dolt_index.go [hunk 1]
```diff
@@ -1230,17 +1230,21 @@ func (di *doltIndex) prollyRangesFromSqlRanges(ctx context.Context, ns tree.Node
...
 			if !field.Hi.Binding || !field.Lo.Binding {
+				// infinity bound
+				fields[i].Unique = false
 				fields[i].Exact = false
 				continue
 			}
-			typ := di.keyBld.Desc.Types[i]
-			cmp := order.CompareValues(i, field.Hi.Value, field.Lo.Value, typ)
-			fields[i].Exact = cmp == 0
+			fields[i].Unique = cmp == 0
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/index/dolt_index.go:1247]: This doesn't seem right, and indicates that these two properties (unique, exact) are being conflated in different circumstances

File: go/libraries/doltcore/sqle/index/dolt_index.go [hunk 2]
```diff
@@ -1230,17 +1230,21 @@ func (di *doltIndex) prollyRangesFromSqlRanges(ctx context.Context, ns tree.Node
...
+			fields[i].Exact = cmp == 0
+
 			if !di.unique {
-				fields[i].Exact = false
+				fields[i].Unique = false
 				continue
 			}
 			if !field.Hi.Binding || !field.Lo.Binding {
+				// infinity bound
+				fields[i].Unique = false
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/index/dolt_index.go:1243]: What does this have to do with uniqueness?

File: go/store/prolly/tuple_range.go [hunk 3]
```diff
@@ -87,7 +90,7 @@ func (r Range) aboveStart(t val.Tuple) bool {
 			return false
 		}
 
-		if r.Fields[i].Exact && cmp == 0 {
+		if r.Fields[i].Unique && cmp == 0 {
```
Comment (zachmu) [inline go/store/prolly/tuple_range.go:93]: This doesn't seem correct to me, semantically -- it has to do with matching semantics, not unique semantics

**Review by zachmu — COMMENTED** [review]
The optimization itself seems fine but I'm not confident the function of the new logic is correct in all cases. Specifically the distinction between unique and exact seems very murky and it's not at all obvious the changes made here work the same as they did before in all cases. I'm worried that we have poor coverage of unique v. non unique index lookups / scans under a variety of comparison operations.

Basically, I don't know that this code is actually correct as far as I can determine by examining it manually, and that bothers me. If you're confident that we actually do have sufficient index text coverage that it demonstrates correctness, then I have no problem with this as written. But you should at least take a second pass through here with an eye to these concerns, some of these changes strike me as sketchy.

**Comment by max-hoffman** [comment]
> The optimization itself seems fine but I'm not confident the function of the new logic is correct in all cases. Specifically the distinction between unique and exact seems very murky and it's not at all obvious the changes made here work the same as they did before in all cases. I'm worried that we have poor coverage of unique v. non unique index lookups / scans under a variety of comparison operations.
> 
> Basically, I don't know that this code is actually correct as far as I can determine by examining it manually, and that bothers me. If you're confident that we actually do have sufficient index text coverage that it demonstrates correctness, then I have no problem with this as written. But you should at least take a second pass through here with an eye to these concerns, some of these changes strike me as sketchy.

I'm reading two concerns, one with testing/correctness and one with interface ambiguity.

I think our testing is pretty solid. We have a ton of tests for indexes, even just in `ComplexIndexQueries`. Sqllogictests also test this pretty extensively. If we changed index analyzer index semantics a lot of things would break loudly. 

I think this works from a correctness standpoint because the analyzer only generates contiguous prefix index scans (it is currently simpler than it used to be), and this PR targets a very specific subset of index scans that should be hard to misidentify.  We transform `range.Matches(partialKey)` to a scan of `[partialKey, partialKey+1)`. Non nil fields defined up to a point, the high is the same as the low for every field, no fields after the first nil are set in the range, and none of the types after the last set are nullable. I think this should reject invalid ranges even if we changed the index generation semantics?

Re: interface, I changed the attribute names to try to be more specific. `.Equality` means the range low and high bounds are equal (for this PR). `.StrictKey` means the range is a strict key into the index and can return 1 or 0 rows (preexisting behavior). Of the two I think `.StrictKey` is probably harder to get right, because of the indirection between `LookupPartitions` and `PartitionRows`. Even for spatial and fulltext indexes, `matches(partialKey)` is still equivalent to `[partialKey, partialKey+1]`.

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>060ce05</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>060ce05</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Review by zachmu — APPROVED** [review]
Talked on the phone, LGTM

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>6c09c4c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>6c09c4c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>fa63080</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>fa63080</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>
